src/streamlink/stream/dash.py

```diff
@@ -0,0 +1,160 @@
+from streamlink.stream.segmented import SegmentedStreamReader, SegmentedStreamWorker, SegmentedStreamWriter
+
+
+class DASHStreamWriter(SegmentedStreamWriter):
+    def __init__(self, reader, *args, **kwargs):
+        options = reader.stream.session.options
+        kwargs["retries"] = options.get("dash-segment-attempts")
+        kwargs["threads"] = options.get("dash-segment-threads")
+        kwargs["timeout"] = options.get("dash-segment-timeout")
+        SegmentedStreamWriter.__init__(self, reader, *args, **kwargs)
+
+    def fetch(self, segment, retries=None):
+        if self.closed or not retries:
+            return
+
+        try:
+            return self.session.http.get(segment.url,
+                                         timeout=self.timeout,
+                                         exception=StreamError)
+        except StreamError as err:
+            self.logger.error("Failed to open segment {0}: {1}", segment.url, err)
+            return self.fetch(segment, retries - 1)
+
+    def write(self, segment, res, chunk_size=8192):
+        for chunk in StreamIOIterWrapper(res.iter_content(chunk_size)):
+            if not self.closed:
+                self.reader.buffer.write(chunk)
+            else:
+                self.logger.debug("Download of segment: {} aborted".format(segment.url))
+                return
+
+        self.logger.debug("Download of segment: {} complete".format(segment.url))
+
+
+class DASHStreamWorker(SegmentedStreamWorker):
+    def __init__(self, *args, **kwargs):
+        SegmentedStreamWorker.__init__(self, *args, **kwargs)
+        self.mpd = self.stream.mpd
+        self.period = self.stream.period
+
+    def iter_segments(self):
+        init = True
+        while not self.closed:
+            # find the representation by ID
+            representation = None
+            for aset in self.mpd.periods[0].adaptionSets:
+                for rep in aset.representations:
+                    if rep.id == self.reader.representation_id:
+                        representation = rep
+            min_wait = self.mpd.minimumUpdatePeriod.total_seconds() if self.mpd.minimumUpdatePeriod else 5
+            with sleeper(min_wait):
+                if representation:
+                    for segment in representation.segments(init=init):
+                        if self.closed:
+                            break
+                        yield segment
+                        self.logger.debug("Adding segment {0} to queue", segment.url)
+
+                    if self.mpd.type == "dynamic":
+                        self.reload()
+                    else:
+                        return
+                    init = False
+
+    def reload(self):
+        if self.closed:
+            return
+
+        self.reader.buffer.wait_free()
+        self.logger.debug("Reloading manifest")
+        res = self.session.http.get(self.mpd.url, exception=StreamError)
+
+        self.mpd = MPD(http.xml(res, ignore_ns=True),
```

Comment:
> AttributeError: 'NoneType' object has no attribute 'xml'

should be `self.session`

Reply:
This was working for me... :)

Reply:
But it is nicer like this, we should aim to remove all the globals :P

Reply:
It should be an issue, don't know why it was working for me :)